README.md
<h1 align="center">Welcome to Habits 👋</h1>

![Licencia MIT](https://img.shields.io/badge/Licencia-MIT-green)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/b3e704c3f150404582cd23b9fcb4be32)](https://www.codacy.com/manual/atareao/habits?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=atareao/habits&amp;utm_campaign=Badge_Grade)
[![CodeFactor](https://www.codefactor.io/repository/github/atareao/habits/badge/master)](https://www.codefactor.io/repository/github/atareao/habits/overview/master)

[![Twitter: atareao](https://img.shields.io/twitter/follow/atareao.svg?style=social)](https://twitter.com/atareao)

Habits is an application to monitor your habits with your PC. So you can study how many kilometers travel with your
mouse or how many keystrokes.

[![Habits](./data/icons/habits.svg)](https://www.atareao.es/aplicacion/habits/)

## 🏠 [Homepage](https://www.atareao.es/aplicacion/habits/)

## Prerequisites

Before you begin, ensure you have met the following requirements:

* If you install it from PPA don't worry about, becouse all the requirements are included in the package
* If you clone the repository, you need, at least, these dependecies,

```
gir1.2-gtk-3.0,
gir1.2-glib-2.0,
gir1.2-gdkpixbuf-2.0,
gir1.2-appindicator3-0.1,
gir1.2-webkit2-4.0,
python3-xlib
```

## Installing Habits

To install **Habits**, follow these steps:

* In a terminal (`Ctrl+Alt+T`), run these commands

```
sudo add-apt-repository ppa:atareao/atareao
sudo apt update
sudo apt install habits
```

## Using Habits

When you start **Habits** it goes to Indicator Area, as you can see in the next screenshot,

![Habits](./screenshots/habits_01.png)

You can set some options in order to start *Habits* when you start your Operative System.

* Set the icon theme. Dark or light.
* Start Habits with your Operative System

Besides you can set the colors of the graphics

![Configuration](./screenshots/habits_02.png)

In the next screenshot you can see the about

![About](./screenshots/habits_03.png)

In the next screenshots you can see some samples of my use of *Habits*,

![Example 1](./screenshots/habits_04.png)

![Example 2](./screenshots/habits_05.png)

## Contributing to Habits

To contribute to **Habits**, follow these steps:

1. Fork this repository.
2. Create a branch: `git checkout -b <branch_name>`.
3. Make your changes and commit them: `git commit -m '<commit_message>'`
4. Push to the original branch: `git push origin <project_name>/<location>`
5. Create the pull request.

Alternatively see the GitHub documentation on [creating a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request).

## 👤 Contributors ✨

Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/docs/en/emoji-key)):

<!-- ALL-CONTRIBUTORS-LIST:START - Do not remove or modify this section -->
<!-- prettier-ignore-start -->
<!-- markdownlint-disable -->
<table>
  <tr>
    <td align="center"><a href="https://www.atareao.es"><img src="https://avatars3.githubusercontent.com/u/298055?v=4" width="100px;" alt=""/><br /><sub><b>Lorenzo Carbonell</b></sub></a><br /><a href="https://github.com/atareao/fondos-productivos/commits?author=atareao" title="Code">💻</a></td>
  </tr>
</table>

<!-- markdownlint-enable -->
<!-- prettier-ignore-end -->
<!-- ALL-CONTRIBUTORS-LIST:END -->

This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!

## Contact

If you want to contact me you can reach me at [atareao.es](https://www.atareao.es).

## License

This project uses the following license: [MIT License](https://choosealicense.com/licenses/mit/).


debian/changelog
habits (0.1.4-0extras19.10.0) bionic; urgency=medium

  * Improve icon resolution

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Thu, 05 Mar 2020 21:34:55 +0100

habits (0.1.3-0extras19.10.0) eoan; urgency=medium

  * Fix a minor typo
  * Add Ducth translation
  * Add units to feet and meters

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Mon, 10 Feb 2020 21:08:03 +0100

habits (0.1.2-0extras19.10.0) eoan; urgency=medium

  * Update Highcharts to version 8.0
  * Add colors preferences to chart

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 28 Dec 2019 21:38:21 +0100

habits (0.1.1-0extras19.10.5) eoan; urgency=medium

  * Add translation to Spanish
  * Change directory
  * Fix minor bugs
  * Remove limit for number of clics
  * Changed color for distance, clics and keys

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Sat, 28 Dec 2019 09:36:44 +0100

habits (0.1.0-0extras19.10.11) eoan; urgency=medium

  * First version

 -- Lorenzo Carbonell <lorenzo.carbonell.cerezo@gmail.com>  Fri, 27 Dec 2019 09:47:50 +0100


src/graph.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of habits
#
# Copyright (c) 2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('WebKit2', '4.0')
except ValueError as e:
    print(e)
    exit(1)
from gi.repository import Gtk
from gi.repository import WebKit2
import config
from basedialog import BaseDialog
from configurator import Configuration


class Graph(BaseDialog):
    def __init__(self, title='', subtitle='', days='', distance='', clics='',
                 keys=''):
        self.title = title
        self.subtitle = subtitle
        self.days = days
        self.distance = distance
        self.clics = clics
        self.keys = keys
        BaseDialog.__init__(self, title, None, ok_button=False,
                            cancel_button=False)

    def init_ui(self):
        BaseDialog.init_ui(self)

        self.scrolledwindow1 = Gtk.ScrolledWindow()
        self.scrolledwindow1.set_policy(Gtk.PolicyType.AUTOMATIC,
                                        Gtk.PolicyType.AUTOMATIC)
        self.grid.attach(self.scrolledwindow1, 0, 0, 1, 1)

        self.viewer = WebKit2.WebView()
        self.scrolledwindow1.add(self.viewer)
        self.scrolledwindow1.set_size_request(900, 600)
        self.viewer.load_uri('file://' + config.HTML_GRAPH)
        self.viewer.connect('load-changed', self.load_changed)
        self.set_focus(self.viewer)

    def update(self):
        configuration = Configuration()
        preferences = configuration.get('preferences')
        units = preferences['units']
        distance = []
        if units == 'feets':
            for i in self.distance:
                distance.append(i/3.28084)
        else:
            distance = self.distance

        self.web_send('title="{}";subtitle="{}";days={};distance={};\
            clics={};keys={};draw_graph(title,subtitle,days,distance,\
                clics, keys);'.format(self.title, self.subtitle, self.days,
                                      distance, self.clics, self.keys))

    def load_changed(self, widget, load_event):
        if load_event == WebKit2.LoadEvent.FINISHED:
            self.update()
            configuration = Configuration()
            preferences = configuration.get('preferences')
            distance_color = preferences['distance-color']
            clics_color = preferences['clics-color']
            keys_color = preferences['keys-color']
            units = preferences['units']
            self.web_send('set_colors("{}", "{}", "{}");'.format(
                distance_color, clics_color, keys_color
            ))
            self.web_send('set_units("{}");'.format(units))
            while Gtk.events_pending():
                Gtk.main_iteration()

    def web_send(self, msg):
        self.viewer.run_javascript(msg, None, None, None)


if __name__ == '__main__':
    title = 'Titulo'
    subtitle = 'Subtitulo'
    days = ['2019-12-25', '2019-12-26', '2019-12-27', 10]
    distance = [25, 30, 35]
    clics = [50, 60, 70]
    keys = [1230, 2550, 2600]
    graph = Graph(title, subtitle, days, distance, clics, keys)
    graph.run()


src/config.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of habits
#
# Copyright (c) 2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import os
import locale
import gettext

PARAMS = {'stats': {},
          'preferences': {'theme-light': True,
                          'start-actived': True,
                          'distance-color': '#445c3c',
                          'clics-color': '#bd574e',
                          'keys-color': '#142d4c',
                          'units': 'meters'}
          }

CONFIG_DIR = os.path.join(os.path.expanduser('~'), '.config/habits')
CONFIG_FILE = os.path.join(CONFIG_DIR, 'habits.conf')


def is_package():
    return __file__.find('src') < 0


APP = 'habits'
APPNAME = 'Habits'

# check if running from source
if is_package():
    ROOTDIR = '/opt/extras.ubuntu.com/habits/share'
    LANGDIR = os.path.join(ROOTDIR, 'locale-langpack')
    APPDIR = os.path.join(ROOTDIR, APP)
    CHANGELOG = os.path.join(APPDIR, 'changelog')
    ICONDIR = os.path.join(ROOTDIR, 'icons')
    HTML_GRAPH = os.path.join(APPDIR, 'graph', 'graph.html')
    AUTOSTARTDIR = os.path.join(ROOTDIR, 'autostart')
else:
    ROOTDIR = os.path.abspath(os.path.dirname(__file__))
    LANGDIR = os.path.normpath(os.path.join(ROOTDIR, '../template1'))
    APPDIR = ROOTDIR
    DEBIANDIR = os.path.normpath(os.path.join(ROOTDIR, '../debian'))
    CHANGELOG = os.path.join(DEBIANDIR, 'changelog')
    ICONDIR = os.path.normpath(os.path.join(ROOTDIR, '../data/icons/'))
    HTML_GRAPH = os.path.join(ROOTDIR, 'graph', 'graph.html')
    AUTOSTARTDIR = os.path.normpath(
        os.path.join(ROOTDIR, '../data/autostart/'))
AUTOSTART = os.path.join(AUTOSTARTDIR, 'habits-autostart.desktop')
ICON = os.path.join(ICONDIR, 'habits.svg')
ICON_ACTIVED_LIGHT = os.path.join(ICONDIR, 'habits-active.svg')
ICON_PAUSED_LIGHT = os.path.join(ICONDIR, 'habits-paused.svg')
ICON_ACTIVED_DARK = os.path.join(ICONDIR, 'habits-active-dark.svg')
ICON_PAUSED_DARK = os.path.join(ICONDIR, 'habits-paused-dark.svg')

f = open(CHANGELOG, 'r')
line = f.readline()
f.close()
pos = line.find('(')
posf = line.find(')', pos)
VERSION = line[pos + 1:posf].strip()
if not is_package():
    VERSION = VERSION + '-src'

try:
    current_locale, encoding = locale.getdefaultlocale()
    language = gettext.translation(APP, LANGDIR, [current_locale])
    language.install()
    print(language)
    if sys.version_info[0] == 3:
        _ = language.gettext
    else:
        _ = language.ugettext
except Exception as e:
    print(e)
    _ = str
APPNAME = _(APPNAME)


src/monitor.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of habits
#
# Copyright (c) 2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import Gtk, Gdk
import sys
import os
import time
import math

from Xlib import X, XK, display
from Xlib.ext import record
from Xlib.protocol import rq
from configurator import Configuration
from threading import Thread


class Monitor(Thread):
    def __init__(self):
        Thread.__init__(self)
        self.daemon = True
        self._running = False
        self.local_dpy = display.Display()
        self.record_dpy = display.Display()
        self.display = Gdk.Display.get_default()
        self.data = {}
        day = time.strftime('%Y-%m-%d', time.localtime())
        configuration = Configuration()
        stats = configuration.get('stats')
        if day in stats:
            self.data[day] = stats[day]

        # Check if the extension is present
        if not self.record_dpy.has_extension("RECORD"):
            print("RECORD extension not found")
            sys.exit(1)
        r = self.record_dpy.record_get_version(0, 0)
        print("RECORD extension version {}.{}".format(r.major_version,
                                                      r.minor_version))
        self.ctx = self.record_dpy.record_create_context(
                0,
                [record.AllClients],
                [{
                        'core_requests': (0, 0),
                        'core_replies': (0, 0),
                        'ext_requests': (0, 0, 0, 0),
                        'ext_replies': (0, 0, 0, 0),
                        'delivered_events': (0, 0),
                        'device_events': (X.KeyPress, X.MotionNotify),
                        'errors': (0, 0),
                        'client_started': False,
                        'client_died': False,
                }])
        default_seat = self.display.get_default_seat()
        _, self.x2, self.y2 = default_seat.get_pointer().get_position()

    def run(self):
        self._running = True
        self.record_dpy.record_enable_context(self.ctx, self.record_callback)

    def stop(self):
        self.local_dpy.record_disable_context(self.ctx)
        self.local_dpy.flush()
        self.record_dpy.record_free_context(self.ctx)
        self._running = False

    def save(self):
        configuration = Configuration()
        stats = configuration.get('stats')
        for day in self.data:
            stats[day] = self.data[day]
        configuration.set('stats', stats)
        configuration.save()

    def lookup_keysym(self, keysym):
        for name in dir(XK):
            if name[:3] == "XK_" and getattr(XK, name) == keysym:
                return name[3:]
        return "[%d]" % keysym

    def record_callback(self, reply):
        if reply.category != record.FromServer:
            return
        if reply.client_swapped:
            print("* received swapped protocol data, cowardly ignored")
            return
        if not len(reply.data) or reply.data[0] < 2:
            return
        data = reply.data
        while len(data):
            event, data = rq.EventField(None).parse_binary_value(
                data, self.record_dpy.display, None, None)
            if event.type == X.KeyPress:
                keysym = self.local_dpy.keycode_to_keysym(event.detail, 0)
                if keysym:
                    # self.inc_data(self.lookup_keysym(keysym), 1)
                    self.inc_data('keys', 1)
            elif event.type == X.ButtonPress:
                # self.inc_data('Button-{}'.format(event.detail), 1)
                self.inc_data('clics', 1)
            elif event.type == X.MotionNotify:
                x1 = self.x2
                y1 = self.y2
                self.x2 = event.root_x
                self.y2 = event.root_y
                monitor2 = self.display.get_monitor_at_point(self.x2, self.y2)
                dx = (self.x2 - x1) * (
                    monitor2.get_width_mm() / monitor2.get_geometry().width)
                dy = (self.y2 - y1) * (
                    monitor2.get_height_mm() / monitor2.get_geometry().height)
                distance = int(math.sqrt(math.pow(dx, 2) + math.pow(dy, 2)))
                self.inc_data('distance', distance)

    def set_data(self, key, value):
        day = time.strftime('%Y-%m-%d', time.localtime())
        if day not in self.data:
            self.data.keys = {}
        self.data[day][key] = value

    def inc_data(self, key, value):
        day = time.strftime('%Y-%m-%d', time.localtime())
        if day not in self.data:
            self.data[day] = {}
        if key not in self.data[day]:
            self.data[day][key] = 0
        self.data[day][key] += value

    def get_data(self, key):
        day = time.strftime('%Y-%m-%d', time.localtime())
        if day in self.data:
            if key in self.data[day].keys:
                return self.data[day][key]
        return 0

    def is_running(self):
        return self._running


if __name__ == '__main__':
    try:
        monitor = Monitor()
        monitor.start()
        Gtk.main()
    except KeyboardInterrupt as e:
        monitor.stop()
        monitor.save()
        print(e)


src/basedialog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of habits
#
# Copyright (c) 2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
except ValueError as e:
    print(e)
    exit(1)
from gi.repository import Gdk, Gtk
import config


class BaseDialog(Gtk.Dialog):
    def __init__(self, title, window=None, ok_button=True, cancel_button=True):
        Gtk.Dialog.__init__(self, title, window)
        self.set_modal(True)
        self.set_destroy_with_parent(True)
        if ok_button:
            self.add_button(Gtk.STOCK_OK, Gtk.ResponseType.ACCEPT)
        if cancel_button:
            self.add_button(Gtk.STOCK_CANCEL, Gtk.ResponseType.CANCEL)
        self.set_default_response(Gtk.ResponseType.ACCEPT)
        self.set_resizable(False)
        self.set_icon_from_file(config.ICON)
        self.connect('realize', self.on_realize)
        self.init_ui()
        self.show_all()

    def init_ui(self):
        vbox0 = Gtk.VBox(spacing=5)
        vbox0.set_border_width(5)
        self.get_content_area().add(vbox0)

        frame1 = Gtk.Frame()
        vbox0.add(frame1)

        self.grid = Gtk.Grid()
        self.grid.set_row_spacing(10)
        self.grid.set_column_spacing(10)
        self.grid.set_margin_bottom(10)
        self.grid.set_margin_start(10)
        self.grid.set_margin_end(10)
        self.grid.set_margin_top(10)
        frame1.add(self.grid)

    def on_realize(self, *_):
        monitor = Gdk.Display.get_primary_monitor(Gdk.Display.get_default())
        scale = monitor.get_scale_factor()
        monitor_width = monitor.get_geometry().width / scale
        monitor_height = monitor.get_geometry().height / scale
        width = self.get_preferred_width()[0]
        height = self.get_preferred_height()[0]
        self.move((monitor_width - width)/2, (monitor_height - height)/2)


if __name__ == '__main__':
    dialog = BaseDialog('Test')
    dialog.run()


src/indicator.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is part of habits
#
# Copyright (c) 2019 Lorenzo Carbonell Cerezo <a.k.a. atareao>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import gi
try:
    gi.require_version('Gtk', '3.0')
    gi.require_version('Gdk', '3.0')
    gi.require_version('AppIndicator3', '0.1')
    gi.require_version('GdkPixbuf', '2.0')
except Exception as e:
    print(e)
    exit(-1)
from gi.repository import Gtk
from gi.repository import Gdk
from gi.repository import GLib
from gi.repository import AppIndicator3
from gi.repository import GdkPixbuf
import webbrowser
from config import _
from monitor import Monitor
from graph import Graph
from preferences import Preferences
import config
from configurator import Configuration


class Indicator(object):

    def __init__(self):

        self.indicator = AppIndicator3.Indicator.new(
            'habits',
            'habits',
            AppIndicator3.IndicatorCategory.APPLICATION_STATUS)
        self.indicator.set_menu(self.build_menu())
        self.indicator.set_label('', '')
        self.indicator.set_status(AppIndicator3.IndicatorStatus.ACTIVE)
        self.monitor = None
        self.load_preferences()
        if self.start_actived:
            self.start()
        else:
            self.stop()
        Gtk.main()

    def set_icon(self, active=True):
        if active:
            if self.theme_light:
                icon = config.ICON_ACTIVED_LIGHT
            else:
                icon = config.ICON_ACTIVED_DARK
        else:
            if self.theme_light:
                icon = config.ICON_PAUSED_LIGHT
            else:
                icon = config.ICON_PAUSED_DARK
        self.indicator.set_icon(icon)

    def load_preferences(self):
        configuration = Configuration()
        preferences = configuration.get('preferences')
        self.theme_light = preferences['theme-light']
        self.start_actived = preferences['start-actived']

    def build_menu(self):
        menu = Gtk.Menu()
        self.menu_toggle_service = Gtk.MenuItem.new_with_label(
            _('Start monitor'))
        self.menu_toggle_service.connect('activate', self.toggle_service)
        menu.append(self.menu_toggle_service)

        menu.append(Gtk.SeparatorMenuItem())

        menu_show_statistics = Gtk.MenuItem.new_with_label(
            _('Statistics'))
        menu_show_statistics.connect('activate', self.show_statistics)
        menu.append(menu_show_statistics)

        menu.append(Gtk.SeparatorMenuItem())

        menu_preferences = Gtk.MenuItem.new_with_label(_('Preferences'))
        menu_preferences.connect('activate', self.show_preferences)
        menu.append(menu_preferences)

        menus_help = Gtk.MenuItem.new_with_label(_('Help'))
        menus_help.set_submenu(self.get_help_menu())
        menu.append(menus_help)

        menu.append(Gtk.SeparatorMenuItem())

        menu_quit = Gtk.MenuItem. new_with_label(_('Quit'))
        menu_quit.connect('activate', self.quit)
        menu.append(menu_quit)
        menu.show_all()
        return menu

    def show_preferences(self, widget):
        widget.set_sensitive(False)
        preferences = Preferences()
        response = preferences.run()
        if response == Gtk.ResponseType.ACCEPT:
            preferences.save()
            self.load_preferences()
            self.set_icon(self.is_monitoring)
        preferences.destroy()
        widget.set_sensitive(True)

    def show_statistics(self, widget):
        widget.set_sensitive(False)

        if self.is_monitoring and self.monitor is not None:
            self.monitor.save()

        title = _('Habits')
        subtitle = _('Mouse and keyboard')
        configuration = Configuration()
        stats = configuration.get('stats')
        days = []
        distance = []
        clics = []
        keys = []
        for day in stats:
            days.append(day)
            if 'distance' in stats[day].keys():
                distance.append(stats[day]['distance']/1000.0)
            else:
                distance.append(0)
            if 'clics' in stats[day].keys():
                clics.append(stats[day]['clics'])
            else:
                clics.append(0)
            if 'keys' in stats[day].keys():
                keys.append(stats[day]['keys'])
            else:
                keys.append(0)
        graph = Graph(title, subtitle, days, distance, clics, keys)
        graph.run()
        graph.destroy()
        widget.set_sensitive(True)

    def get_help_menu(self):
        help_menu = Gtk.Menu()

        homepage_item = Gtk.MenuItem.new_with_label(_('Homepage'))
        homepage_item.connect(
            'activate',
            lambda x: webbrowser.open('http://www.atareao.es/apps/habits/'))
        help_menu.append(homepage_item)

        help_item = Gtk.MenuItem.new_with_label(_('Get help online...'))
        help_item.connect(
            'activate',
            lambda x: webbrowser.open('http://www.atareao.es/apps/habits/'))
        help_menu.append(help_item)

        translate_item = Gtk.MenuItem.new_with_label(_(
            'Translate this application...'))
        translate_item.connect(
            'activate',
            lambda x: webbrowser.open('http://www.atareao.es/apps/habits/'))
        help_menu.append(translate_item)

        bug_item = Gtk.MenuItem.new_with_label(_('Report a bug...'))
        bug_item.connect(
            'activate',
            lambda x: webbrowser.open('https://github.com/atareao\
/habits/issues'))
        help_menu.append(bug_item)

        help_menu.append(Gtk.SeparatorMenuItem())

        twitter_item = Gtk.MenuItem.new_with_label(_('Found me in Twitter'))
        twitter_item.connect(
            'activate',
            lambda x: webbrowser.open('https://twitter.com/atareao'))
        help_menu.append(twitter_item)
        #
        github_item = Gtk.MenuItem.new_with_label(_('Found me in GitHub'))
        github_item.connect(
            'activate',
            lambda x: webbrowser.open('https://github.com/atareao'))
        help_menu.append(github_item)

        mastodon_item = Gtk.MenuItem.new_with_label(_('Found me in Mastodon'))
        mastodon_item.connect(
            'activate',
            lambda x: webbrowser.open('https://mastodon.social/@atareao'))
        help_menu.append(mastodon_item)

        about_item = Gtk.MenuItem.new_with_label(_('About'))
        about_item.connect('activate', self.menu_about_response)

        help_menu.append(Gtk.SeparatorMenuItem())

        help_menu.append(about_item)
        return help_menu

    def menu_about_response(self, widget):
        widget.set_sensitive(False)
        ad = Gtk.AboutDialog()
        ad.set_name(config.APPNAME)
        ad.set_version(config.VERSION)
        ad.set_copyright('Copyrignt (c) 2019\nLorenzo Carbonell')
        ad.set_comments(_('Habits'))
        ad.set_license('''
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:
The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.''')
        ad.set_website('')
        ad.set_website_label('http://www.atareao.es')
        ad.set_authors(['Lorenzo Carbonell Cerezo <a.k.a. atareao>'])
        ad.set_translator_credits('Lorenzo Carbonell Cerezo <a.k.a. atareao>')
        ad.set_documenters(['Lorenzo Carbonell Cerezo <a.k.a. atareao>'])
        ad.set_artists(['Freepik <https://www.flaticon.com/authors/freepik>'])
        ad.set_logo(GdkPixbuf.Pixbuf.new_from_file(config.ICON))
        ad.set_icon(GdkPixbuf.Pixbuf.new_from_file(config.ICON))
        ad.set_program_name(config.APPNAME)

        monitor = Gdk.Display.get_primary_monitor(Gdk.Display.get_default())
        scale = monitor.get_scale_factor()
        monitor_width = monitor.get_geometry().width / scale
        monitor_height = monitor.get_geometry().height / scale
        width = ad.get_preferred_width()[0]
        height = ad.get_preferred_height()[0]
        ad.move((monitor_width - width)/2, (monitor_height - height)/2)

        ad.run()
        ad.destroy()
        widget.set_sensitive(True)

    def toggle_service(self, widget):
        if self.monitor is None:
            self.start()
        else:
            self.stop()

    def stop(self):
        self.set_icon(False)
        self.is_monitoring = False

        self.menu_toggle_service.set_label(_('Start monitor'))
        self.monitor.stop()
        self.monitor.save()
        self.monitor = None

    def start(self):
        self.set_icon(True)
        self.is_monitoring = True

        self.menu_toggle_service.set_label(_('Stop monitor'))
        self.monitor = Monitor()
        self.monitor.start()

    def quit(self, menu_item):
        if self.monitor is not None:
            self.stop()
        Gtk.main_quit()
        # If Gtk throws an error or just a warning, main_quit() might not
        # actually close the app
        sys.exit(0)


def main():
    Indicator()


if __name__ == '__main__':
    main()
